executor/src/witgen/expression_evaluator.rs

```diff
@@ -43,9 +49,8 @@ where
                 self.evaluate_binary_operation(left, op, right)
             }
             Expression::UnaryOperation(op, expr) => self.evaluate_unary_operation(op, expr),
-            e => Err(IncompleteCause::ExpressionEvaluationUnimplemented(
-                e.to_string(),
-            )),
+            Expression::Challenge(challenge) => self.variables.challenge(challenge),
+            e => unimplemented!("Unexpected expression: {}", e),
```

Comment:
I think this should be a hard failure, not an "incomplete" result.

Reply:
It might be that some factor will be zero later, and thus we do not need to evaluate this branch, but I guess the expression is still mal-formed, so fine with hard error.